Commercial skipping notifications and toggle action

b-pass opened the pull request:
An implementation to address this feature request from the forums: http://forum.kodi.tv/showthread.php?tid=208206

This adds a new action, TOGGLE_COMMSKIP which turns Commercial Skipping on/off for the current recording only (next record starts back at the default setting from advancedsettings.xml).

Also adds an option to display notifications of commercials and their lengths.

This is a resubmit of #7330 and #7520.

b-pass committed:
Commercial skipping notifications and toggle action

FernetMenta commented on xbmc/video/windows/GUIWindowFullScreen.cpp:
```diff
@@ -177,6 +177,11 @@ bool CGUIWindowFullScreen::OnAction(const CAction &action)
     }
     break;
 
+  case ACTION_TOGGLE_COMMSKIP:
```
player can handle actions by itself. no need to implement a new method to the interface

FernetMenta commented on xbmc/settings/AdvancedSettings.h:
```diff
@@ -287,6 +287,8 @@ class CAdvancedSettings : public ISettingCallback, public ISettingsHandler
     int m_iEdlMaxStartGap;          // seconds
     int m_iEdlCommBreakAutowait;    // seconds
     int m_iEdlCommBreakAutowind;    // seconds
+    bool m_bEdlCommAutoSkip; /*!< @brief true to automatically skip commercial breaks in Edit Decision Lists by defeault */
```
this is dead code. the average user is not aware of advanced settings. please drop.

b-pass commented:
@FernetMenta this was requested to be in advanced settings by @ksooo, @MartijnKaijser, and you in #7330 because everyone was worried it would be a support problem if it showed up in a more prominent place.

FernetMenta commented:
@b-pass I can't find any reference that I requested this to be in advancedsettings. I only found a single post of myself in this thread:

> support for comskip is no requirement for a pvr addon. hence adding those options to general settings is not correct. this needs to be enabled by something private to an addon and conveyed through the pvr api.

what sources do require this feature?

b-pass commented:
From @ksooo 
> @MartijnKaijser yes, I'm actually afraid of this "support nightmare", too. We have to be really very clear about what this power user feature is/is not if we expose it through any UI.

Sounds to me like the point was to keep it out of the GUI entirely.  It's an EDL feature, which is just sprinkled throughout VideoPlayer and already itself has a few settings in Advanced settings. In #7330 there's some discussion of a "support nightmare" if the setting is left in the GUI anywhere.

FernetMenta commented:
Why don't you disable the function by default and have users enable it with the toggle on demand? THat would eliminate the need for a setting.

b-pass commented:
@FernetMenta That's fair.  The setting was to allow people to turn it off. But since it can't be turned off today either, there's not really a loss in functionality.  I've made the changes you requested.

FernetMenta commented:
you still have ToggleCommSkip in the interface. This is not required because VideoPlayer listens to action event.

FernetMenta commented on xbmc/cores/VideoPlayer/VideoPlayer.cpp:
```diff
@@ -4391,6 +4418,9 @@ bool CVideoPlayer::OnAction(const CAction &action)
       else
         break;
+    case ACTION_TOGGLE_COMMSKIP:
+      ToggleCommSkip();
```
ahh, I think it is better to place those two lines of code here instead of call a function.

BrettEBowman replied:
> Why don't you disable the function by default and have users enable it with the toggle on demand? THat would eliminate the need for a setting.

Disabling the function by default **changes the current behavior** of the player which has been using the EDL files to cut, mute, or skip over indicated sections of all files being played for quite some time now. As mentioned in another message, this functionality pre-dates PVR functionality in Kodi/XBMC. The original reason for adding a toggle to disable this function is that occasionally the EDL files which are generated are "corrupt." (Technically, they aren't corrupt since they are valid files, they just have very wrong time points for these cuts/mutes/skips). In those cases, it is desirable to **temporarily disable** the automatic actions for that single, specific player session.

I don't see why adding one more EDL configuration option to the existing set of them in the advancedsettings is a problem. But if that must be avoided, then the default should be for EDL processing to be enabled so that the current functionality of the player is not changed.

b-pass commented:
@BrettEBowman I've done as you said, and EDL will continue to be enabled by default, the action can turn it off.  The setting which I've now removed was to allow it to be disabled by default.

@FernetMenta I've removed the unnecessary 2 line function now as well.

FernetMenta commented:
jenkins build this please

FernetMenta commented:
thanks!

enen92 commented:
@jpribyl see https://github.com/xbmc/xbmc/pull/20711 (Kodi 20 and up)

enen92 commented:
No, unfortunately not. That's sort of on my todo list but we still lack any way of checking from the skin engine if we "come" from an edl mark after a seek.
At the moment you'll have to patch estuary to avoid showing the seek bar. That can be done removing `Player.DisplayAfterSeek` calls for < 20 or removing `Player.HasPerformedSeek(x)` calls for v20 and up.
I'll likely come back to this at some point after v20 is out.

b-pass commented:
If you don't want to see the progress bar or the notification then you should be setting your EDL lists to "Cut" type instead of "CommBreak" type.  That's the whole purpose of "Cut" type -- to make it seem like the skipped section was not there at all (skip as smoothly as possible, etc).

How exactly you do that depends on the source and/or file format of the EDL.

enen92 commented:
Cut effectively removes the block from the timeline, you're not allowed to jump back into the removed section if for some reason the software marks the advertisement on an incorrect position (quite common). Also, internally the player does a seek anyway so the seekbar will also display..

calania commented:
Am really interested in the TOGGLE_COMMSKIP function but can't I can't find how to use it. Could anybody here please explain how it works?

b-pass commented:
> Am really interested in the TOGGLE_COMMSKIP function but can't I can't find how to use it. Could anybody here please explain how it works?

Yes, you can reference it in a keymap xml file.  For example, to have the button named "green" do this action while watching video:

`<keymap><fullscreenvideo><remote><green>ToggleCommSkip</green></remote></fullscreenvideo>`

calania commented:
Didn't expect the original creator of this pull request to answer after so many years :) . Thanks for the answer but it's not exactly what I want. I would want it to by default not skip the commercials and having to always press a remote on to toggle auto skip off is quite annoying. In an ideal scenario I would want a popup along the lines of "Commercial detected! Do you want to skip?" but simply disabling auto skip would work as well. Looking through your pull requests it seems like there once was a way to disable auto skip in the advancedsettings.xml so it didn't skip by default (you even mentioned it in the first post of this pr). But I cant find any mention of it now. Did you or someone else remove the ability to set the default behavior?

b-pass commented:
> Looking through your pull requests it seems like there once was a way to disable auto skip in the advancedsettings.xml so it didn't skip by default (you even mentioned it in the first post of this pr). But I cant find any mention of it now. Did you or someone else remove the ability to set the default behavior?

At the time, they did not want to add things to `advancedsettings.xml`, but they also did not want to add options to the main Kodi settings pages which implied commercial skipping could be done by kodi alone.  So there isn't a way to configure it, because there didn't seem to be something that the PR approvers would agree on.

calania commented:
Well that's a shame. Seems like quite a basic setting and can't really understand why at least not add it to advancedsettings.xml
